### PR TITLE
feat(openapi): crete an upgrader that transitions api.categories from…

### PR DESCRIPTION
**Pull request by illichso**
… a list of keys to a list of ids

## Issue

https://gravitee.atlassian.net/browse/APIM-4851

## Description

A small description of what you did in that PR.

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-nbqtzlkfiy.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+        when(categoryRepository.findAllByEnvironment("env1")).thenReturn(Set.of(category1, category2));
+
+        apiV4CategoriesUpgrader.upgrade();
+
+        verify(apiRepository, times(1)).update(any());
+    }
+
+    @Test
+    public void shouldNotMigrateV4ApiCategoriesWhenNoCategoriesExist() throws Exception {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]:
```suggestion
    public void shouldNotMigrateV4ApiCategoriesWhenNoV4ApisWithCategoriesExist() throws Exception {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        api.setCategories(new HashSet<>(Arrays.asList("category1", "category2")));
+        api.setEnvironmentId("env1");
+        api.setDefinitionVersion(DefinitionVersion.V4);
+
+        when(apiRepository.search(any(), any(), any())).thenReturn(Stream.of(api));
+        when(categoryRepository.findAllByEnvironment("env1")).thenReturn(Set.of());
+
+        apiV4CategoriesUpgrader.upgrade();
+
+        verify(apiRepository, times(0)).update(any());
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]:
```suggestion
        verify(apiRepository, never()).update(any());
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        Category category2 = new Category();
+        category2.setKey("category2");
+        category2.setId("id2");
+        category2.setEnvironmentId("env1");
+
+        when(categoryRepository.findAllByEnvironment("env1")).thenReturn(Set.of(category1, category2));
+
+        apiV4CategoriesUpgrader.upgrade();
+
+        verify(apiRepository, times(1)).update(any());
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]: Could you check that the saved API contains ids in the list?
Something like this:
```suggestion
        verify(apiRepository, times(1))
           .update(argThat(api -> api.categories.containsAll(List.of("id1", "id2")));
```

**Review by phiz71 — CHANGES_REQUESTED** [review]
LGTM ! Good job !
I would like you to add a last check in a test (+ non important changes)

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java (hunk 2 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java (hunk 3 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java (hunk 1 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/ApiV4CategoriesUpgraderTest.java]: Done

**Review by phiz71 — APPROVED** [review]

**Pull request merged**